### PR TITLE
Hazelcast 4 - another try

### PR DESCRIPTION
**Pull request by leszko**
Guys, please check, some other idea how we could address the compatibility issue.

Please check: @mesutcelik @pivovarit @alparslanavci

### CODE REVIEW

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/HazelcastAdapter.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+    private static Class nearCacheStatsClass() throws ClassNotFoundException {
+        try {
+            return Class.forName("com.hazelcast.nearcache.NearCacheStats");
+        } catch (ClassNotFoundException e) {
+            return Class.forName("com.hazelcast.monitor.NearCacheStats");
+        }
+    }
+
+    static class IMap<K, V> {
```
Comment (pivovarit) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/HazelcastAdapter.java]: This implementation could just implement the original IMap

Comment (alparslanavci) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/HazelcastAdapter.java]: if it implements the original IMap, then doesn't it need the IMap class at compile time?

Comment (pivovarit) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/HazelcastAdapter.java]: 🤦‍♀️

Comment (pivovarit) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/HazelcastAdapter.java]: For a second I thought we could avoid dropping `C extends IMap<K, V>` generic bounds, but yeah - that's a no-go

File: dependencies.gradle [hunk 2]
```diff
@@ -11,7 +11,7 @@ def VERSIONS = [
         'com.google.dagger:dagger-compiler:2.11',
         'com.google.guava:guava:latest.release',
         'com.google.inject:guice:4.1.0',
-        'com.hazelcast:hazelcast:3.+',
+        'com.hazelcast:hazelcast:4.+',
```
Comment (pivovarit) [inline dependencies.gradle:14]: We should actually keep it `3.+` by default I believe

**Comment by mesutcelik** [comment]
Can you try to upgrade spring boot hazelcast dependency to hazelcast 4 along with a micrometer dependency including this fix? If we are able to build spring boot from source with hazelcast 4 dependency then we can present a working solution to spring boot team.

**Pull request closed without merging**